File: gtci/03-sliding-window/python3/1-maximum-sum-subarray-of-size-k.py
class Solution:
    def findMaxSumSubArray(self, K, arr):
        max_sum = float("-inf")

        window_start = 0
        window_sum = 0

        for window_end in range(len(arr)):
            window_sum += arr[window_end]

            if window_end >= K - 1:
                max_sum = max(max_sum, window_sum)

                window_sum -= arr[window_start]
                window_start += 1

        return max_sum


if __name__ == "__main__":
    sol = Solution()

    assert sol.findMaxSumSubArray(3, [2, 1, 5, 1, 3, 2]) == 9
    assert sol.findMaxSumSubArray(2, [2, 3, 4, 1, 5]) == 7

    print("All test cases passed.")

File: gtci/03-sliding-window/cpp/1-maximum-sum-subarray-of-size-k.cpp
#include <cassert>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

class Solution {
   public:
    int findMaxSumSubArray(const int K, const vector<int>& arr) {
        int maxSum = std::numeric_limits<int>::min();

        int windowStart = 0;
        int windowSum = 0;

        for (int windowEnd = 0; windowEnd < arr.size(); windowEnd++) {
            windowSum += arr[windowEnd];

            if (windowEnd >= K - 1) {
                maxSum = std::max(maxSum, windowSum);

                windowSum -= arr[windowStart];
                windowStart++;
            }
        }

        return maxSum;
    }
};

int main() {
    Solution sol;

    assert(sol.findMaxSumSubArray(3, {2, 1, 5, 1, 3, 2}) == 9);
    assert(sol.findMaxSumSubArray(2, {2, 3, 4, 1, 5}) == 7);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}
